Mathlib/Tactic/Pushneg.lean

```diff
@@ -0,0 +1,119 @@
+/-
+Copyright (c) 2022 Alice Laroche. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Alice Laroche
+-/
+
+import Lean
+import Lean.Meta
+import Lean.Elab
+import Mathlib.Lean.Expr
+import Mathlib.Logic.Basic
+
+namespace Mathlib.Tactic
+
+open Lean Meta Elab.Tactic Parser.Tactic
+open Lean.Expr
+open Lean.Elab.Term
+
+theorem not_not_eq (h : p = p') : (¬¬p) = p' := h ▸ (propext not_not)
+theorem not_and_eq (h : p = p')    (h' : (¬q) = q') : (¬ (p ∧ q)) = (p' → q') := h ▸ h' ▸ propext not_and
+theorem not_or_eq  (h : (¬p) = p') (h' : (¬q) = q') : (¬ (p ∨ q)) = (p' ∧ q') := h ▸ h' ▸ propext not_or_distrib
+theorem not_forall_eq {s s' : α → Prop} (h : ∀x, (¬s x) = s' x) : (¬ ∀x, s x) = (∃ x, s' x) := funext h ▸ propext not_forall
+theorem not_exists_eq {s s' : α → Prop} (h : ∀x, (¬s x) = s' x) : (¬ ∃x, s x) = (∀ x, s' x) := forall_congr h ▸ propext not_exists
+
+def binderTelescope1 (e : Expr) (f : Expr → Expr → MetaM α) : MetaM α := do
+match e with
+| Expr.lam n t b d => withLocalDecl n d.binderInfo t fun x => f x (instantiate1 b x)
+| Expr.forallE n t b d => withLocalDecl n d.binderInfo t fun x => f x (instantiate1 b x)
+| _ => throwError "binder expected{indentExpr e}"
+
+partial def pushNegation (expr : Expr) : MetaM (Expr × Expr) := do
+match expr with
+| Expr.forallE n t b _ =>
```

Comment:
This doesn't handle implication, that is, I think it would fail for `¬ (a ∧ b) → c`

Reply:
The handling of such expression will occur in the `Expr.app` case, as it begin by a negation.
However you're right, it will treat the implication as a normal forall, so it will be transformed into an existantial. I will change that.

Reply:
Ah right, i forget there could be a negation in the domain. I will change that here and for all binder/quantifier.

Reply:
This also remind that implication must be transformed into and, not into existential binder.